// grammy/out/convenience/frameworks.web.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.adapters = void 0;
// deno-lint-ignore-file no-explicit-any
const frameworks_shared_js_1 = require("./frameworks.shared.js");
/** Native CloudFlare workers (service worker) */
const cloudflare = (event) => {
    let resolveResponse;
    event.respondWith(new Promise((resolve) => {
        resolveResponse = resolve;
    }));
    return {
        update: event.request.json(),
        header: event.request.headers.get(frameworks_shared_js_1.SECRET_HEADER) || undefined,
        end: () => {
            resolveResponse(new Response(null, { status: 200 }));
        },
        respond: (json) => {
            const res = new Response(json, {
                status: 200,
                headers: { "Content-Type": "application/json" },
            });
            resolveResponse(res);
        },
        unauthorized: () => {
            const res = new Response('"unauthorized"', {
                status: 401,
                statusText: "secret token is wrong",
            });
            resolveResponse(res);
        },
    };
};
/** Native CloudFlare workers (module worker) */
const cloudflareModule = (request) => {
    let resolveResponse;
    return {
        update: request.json(),
        header: request.headers.get(frameworks_shared_js_1.SECRET_HEADER) || undefined,
        end: () => {
            resolveResponse(new Response(null, { status: 200 }));
        },
        respond: (json) => {
            const res = new Response(json, {
                status: 200,
                headers: { "Content-Type": "application/json" },
            });
            resolveResponse(res);
        },
        unauthorized: () => {
            const res = new Response('"unauthorized"', {
                status: 401,
                statusText: "secret token is wrong",
            });
            resolveResponse(res);
        },
        handlerReturn: new Promise((resolve) => {
            resolveResponse = resolve;
        }),
    };
};
/** hono web framework */
const hono = (ctx) => {
    let resolveResponse;
    return {
        update: ctx.req.json(),
        header: ctx.req.headers.get(frameworks_shared_js_1.SECRET_HEADER) || undefined,
        end: () => {
            resolveResponse(ctx.body());
        },
        respond: (json) => {
            ctx.header('Content-Type", "application/json');
            resolveResponse(ctx.body(json));
        },
        unauthorized: () => {
            ctx.status(401);
            ctx.statusText("secret token is wrong");
            resolveResponse(ctx.body());
        },
        handlerReturn: new Promise((resolve) => {
            resolveResponse = resolve;
        }),
    };
};
/** worktop CloudFlare workers framework */
const worktop = (req, res) => ({
    update: Promise.resolve(req.body.json()),
    header: req.headers.get(frameworks_shared_js_1.SECRET_HEADER),
    end: () => res.end(),
    respond: (json) => res.send(200, json),
    unauthorized: () => res.send(401, "secret token is wrong"),
});
exports.adapters = {
    cloudflare,
    "cloudflare-mod": cloudflareModule,
    hono,
    worktop,
};
